### ISSUE TITLE
nbgrader generating SQL errors around the use of the `sum` statement

### ISSUE THREAD

**Issue by Wildcarde** [jupyter/nbgrader#1246]
### Operating system
rhel 7, but currently running in a base-notebook container in docker
### `nbgrader --version`
0.5.5-0.6.0
backended by an sqlite3 database
### `jupyter notebook --version` 
6.0.0

### Expected behavior
show list of students
### Actual behavior
web interface appears to poll forever.  On the logging side you get:
```
test_1  | [E 20:03:00.407 NotebookApp] Uncaught exception GET /formgrader/api/students (172.18.0.1)
test_1  |     HTTPServerRequest(protocol='http', host='localhost:9999', method='GET', uri='/formgrader/api/students', version='HTTP/1.1', remote_ip='172.18.0.1')
test_1  |     Traceback (most recent call last):
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1244, in _execute_context
test_1  |         cursor, statement, parameters, context
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 550, in do_execute
test_1  |         cursor.execute(statement, parameters)
test_1  |     sqlite3.OperationalError: misuse of aggregate: sum()
test_1  |
test_1  |     The above exception was the direct cause of the following exception:
test_1  |
test_1  |     Traceback (most recent call last):
test_1  |       File "/opt/conda/lib/python3.7/site-packages/tornado/web.py", line 1697, in _execute
test_1  |         result = method(*self.path_args, **self.path_kwargs)
test_1  |       File "/opt/conda/lib/python3.7/site-packages/tornado/web.py", line 3174, in wrapper
test_1  |         return method(self, *args, **kwargs)
test_1  |       File "/opt/conda/lib/python3.7/site-packages/nbgrader/server_extensions/formgrader/base.py", line 103, in wrapper
test_1  |         return f(self, *args, **kwargs)
test_1  |       File "/opt/conda/lib/python3.7/site-packages/nbgrader/server_extensions/formgrader/apihandlers.py", line 172, in get
test_1  |         students = self.api.get_students()
test_1  |       File "/opt/conda/lib/python3.7/site-packages/nbgrader/apps/api.py", line 726, in get_students
test_1  |         students = gb.student_dicts()
test_1  |       File "/opt/conda/lib/python3.7/site-packages/nbgrader/api.py", line 2321, in student_dicts
test_1  |         total_score, = self.db.query(func.sum(Assignment.max_score)).one()
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3282, in one
test_1  |         ret = self.one_or_none()
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3251, in one_or_none
test_1  |         ret = list(self)
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3324, in __iter__
test_1  |         return self._execute_and_instances(context)
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3349, in _execute_and_instances
test_1  |         result = conn.execute(querycontext.statement, self._params)
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 988, in execute
test_1  |         return meth(self, multiparams, params)
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 287, in _execute_on_connection
test_1  |         return connection._execute_clauseelement(self, multiparams, params)
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1107, in _execute_clauseelement
test_1  |         distilled_params,
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1248, in _execute_context
test_1  |         e, statement, parameters, cursor, context
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1466, in _handle_dbapi_exception
test_1  |         util.raise_from_cause(sqlalchemy_exception, exc_info)
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 399, in raise_from_cause
test_1  |         reraise(type(exception), exception, tb=exc_tb, cause=cause)
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 153, in reraise
test_1  |         raise value.with_traceback(tb)
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1244, in _execute_context
test_1  |         cursor, statement, parameters, context
test_1  |       File "/opt/conda/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 550, in do_execute
test_1  |         cursor.execute(statement, parameters)
test_1  |     sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) misuse of aggregate: sum()
test_1  |     [SQL: SELECT sum((SELECT coalesce(sum(grade_cell.max_score), ?) AS coalesce_1
test_1  |     FROM grade_cell, notebook, assignment
test_1  |     WHERE notebook.assignment_id = assignment.id AND grade_cell.notebook_id = notebook.id)) AS sum_1]
test_1  |     [parameters: (0.0,)]
test_1  |     (Background on this error at: http://sqlalche.me/e/e3q8)
test_1  | [E 20:03:00.417 NotebookApp] {
test_1  |       "Host": "localhost:9999",
test_1  |       "Connection": "keep-alive",
test_1  |       "Accept": "application/json, text/javascript, */*; q=0.01",
test_1  |       "Dnt": "1",
test_1  |       "X-Csrftoken": "2|f47d4a5a|9b80fe8a45361d4d00525861e5ca33b8|1571266187",
test_1  |       "X-Requested-With": "XMLHttpRequest",
test_1  |       "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.120 Safari/537.36",
test_1  |       "Sec-Fetch-Mode": "cors",
test_1  |       "Sec-Fetch-Site": "same-origin",
test_1  |       "Referer": "http://localhost:9999/formgrader/manage_students",
test_1  |       "Accept-Encoding": "gzip, deflate, br",
test_1  |       "Accept-Language": "en-US,en;q=0.9",
test_1  |       "Cookie": "_xsrf=2|f47d4a5a|9b80fe8a45361d4d00525861e5ca33b8|1571266187; username-localhost-9999=\"2|1:0|10:1571342504|23:username-localhost-9999|44:NjA0ZGQ1ZTFkZDZkNDEzZDkzMzBhNmRkZmIyMjBjNjQ=|787ce66caf41fd79f5c6d146e5ce0e029b2b473f86d6070c8a98bcbe794f4332\""
test_1  |     }
test_1  | [E 20:03:00.417 NotebookApp] 500 GET /formgrader/api/students (172.18.0.1) 643.93ms referer=http://localhost:9999/formgrader/manage_students
```

### Steps to reproduce the behavior
build a base notebook container mounting all the relevant pieces for the exchange folder, database, and related bits and pieces. Attempt to list students inside the container.

test docker file built using our jupyterhub container (there's a lot of noise here):
```
FROM jupyter/base-notebook:latest

## Install dependencies
#RUN apt-get -y update && \
#    apt-get -y upgrade && \
 #   apt-get -y install krb5-user sssd git vim && \
 #   apt-get purge && \
 #   apt-get clean && \
 #   rm -rf /var/lib/apt/lists/* && \
 #   git clone https://github.com/jupyterhub/dockerspawner.git /tmp/dockerspawner

#ADD dockerspawner /tmp/dockerspawner
USER root

RUN pip install --no-cache jupyter-client notebook nbgrader scikit-learn  && \
     jupyter nbextension enable nbgrader --py --sys-prefix && \
    jupyter nbextension install --sys-prefix --py nbgrader --overwrite && \
    jupyter nbextension enable --sys-prefix --py nbgrader && \
    jupyter serverextension enable --sys-prefix --py nbgrader

RUN conda update --all && conda update -n base conda && \
    conda install -c conda-forge xorg-libxrender xorg-libxpm nbgrader matplotcheck -y && \
    conda install jupyterhub=0.9.4 nose invoke termcolor pandas scikit-learn  -y && \
    /opt/conda/bin/pip install plotchecker


## install repaired swarmspawner tooling
##COPY swarmspawnergdb /opt/swarmspawnergdb

##RUN cd /opt/swarmspawnergdb && pip install . && /opt/conda/bin/pip install .


VOLUME ['/mnt/classdat']
VOLUME ['/mnt/exchange']
VOLUME ['/mnt/db'] ## location of sqlite database for classes.
## setup local user for jhub grader system, jovyan is uid 1000, this will be 1999
#to keep them rationally seperate

RUN groupadd -g 1001 psy360student && groupadd -g 1002 psy360instructor

RUN groupadd -g 1999 s_nbgrader && \
    useradd -u 1999 s_nbgrader -g s_nbgrader -s /bin/bash && \
    mkdir -p /home/s_nbgrader/servicewd && \
    chown -Rv 1999:1999 /home/s_nbgrader && \
    usermod -a -G psy360student s_nbgrader && \
    usermod -a -G psy360instructor s_nbgrader


#COPY jupyterhub_config.py /srv/jupyterhub/
COPY nbgrader_config.central.py /etc/jupyter/nbgrader_config.py
COPY nbgrader_config.serviceaccount.py /home/s_nbgrader/servicewd/nbgrader_config.py
#COPY cull_idle_servers.py /srv/jupyterhub/
RUN  chown -Rv 1999:1999 /home/s_nbgrader/*
RUN chown -Rv 1999:1999 /home/jovyan



WORKDIR "/mnt/classdat"

RUN conda install sqlalchemy==1.3 nbgrader==0.5.5 -y && pip install sqlalchemy==1.3 nbgrader==0.5.5

#RUN conda install nbgrader==0.5.6 -y && pip install nbgrader==0.5.6

USER s_nbgrader

RUN jupyter nbextension enable nbgrader --py --user && jupyter serverextension enable nbgrader
```

compose file:
```
version: "3"
services:
  test:
    # build an image with SwarmSpawner and our jupyterhub_config.py
    #env_file: .env
    build:
      context: "."
      dockerfile: "test.Dockerfile"
    # mount the docker socket
    volumes:
      - "/mnt/xxxx/home/:/mnt/lockhart/PSY255/home/:rw"
      - "/opt/nbgrader-test/data/:/mnt/db:rw" ## we use relative paths to make sure the database file is stored on local disk instead of an NFS mount to avoid some known sqlite+nfs issues.
      - "/mnt/xxxx/exchange:/mnt/exchange"
      - "/mnt/xxxx/classdat:/mnt/classdat"
    ports:
      - 9999:8888
```

**Comment by Wildcarde** [comment]
I should note, this worked as expected in august and broke when we regenerated the image this past week. As such i've tested all versions of nbgrader from this summer till now and am currently working my way through sqlalchemy versions in an attempt to find where the failure gets imported.

**Comment by perllaghu** [comment]
> working my way through sqlalchemy versions in an attempt to find where the failure gets imported.

There was a problem with SQLAlchemy when it went from 1.2.8 to 1.3.... I can't remember if that got fixed :(

**Comment by Wildcarde** [comment]
I'm currently building a container that runs python 3.6 instead of 3.7 so I can test older versions of sqlalchemy in order to verify this in particular.

**Comment by Wildcarde** [comment]
Looking at the list of available files for sqlalchemy, is it possible this was introduced going from 1.2.7 to 1.2.8? 1.2.8 appears to be the first python 3.7 compatible version release on anaconda.

**Comment by Wildcarde** [comment]
It does appear python 3.6 + sqlalchemy 1.2.7 + nbgrader 0.6.0 is able to list out students and grades unlike python 3.7 and any newer versions of sqlalchemy.

**Comment by Wildcarde** [comment]
Additional critical note, SQLite 3.30 does not work even with the above combination of packages. Using 3.28 appears to but I'm still working on some testing.

**Comment by perllaghu** [comment]
@Wildcarde - can I just thank you for doing all this testing!

**Comment by Wildcarde** [comment]
happy to help, we are kinda stuck with a class waiting for grades till this all gets squared on our end. 
I've started working on a much more stripped down container based on jupyterhub/jupyterhub:0.9.4 and the following three lines will result in a functional system on my local test container:
```
RUN conda update --all && conda update -n base conda

RUN conda install nose invoke \
    termcolor pandas scikit-learn sqlite==3.28 sqlalchemy==1.3.9 -y && \
    conda install -c conda-forge nbgrader==0.6.0 xorg-libxrender xorg-libxpm matplotcheck -y && \
    /opt/conda/bin/pip install /tmp/dockerspawner plotchecker

RUN jupyter nbextension install --sys-prefix --py nbgrader --overwrite && \
    jupyter nbextension enable --sys-prefix --py nbgrader && \
    jupyter serverextension enable --sys-prefix --py nbgrader

```

Turns out the critical bit here (currently) is sqlite as I'm running the latest versions of sqlalchemy and nbgrader. I'm currently adapting these changes to our classroom jupyterhub instance and will hopefully be rolling the fixed setup into production soon.

**Comment by nthiery** [comment]
Got hit by this too. Thanks so much @Wildcarde for the tip; setting the fol my environment.yml file saved my day!
```
- sqlite==3.28
- sqlalchemy==1.3.9
```

**Comment by jhamrick** [comment]
Thanks for tracking this down!

I did a bit of digging to try to fix and think I've come up with a solution. The offending query is:

```
self.db.query(func.sum(Assignment.max_score)).one()
```

I've submitted #1280 which seems to at least fix the problem for me locally.

**Comment by jhamrick** [comment]
Update: I can't seem to easily test this in the CI tests since apt ships with an older version of sqlite. I'll run all the tests locally and see if any others fail as well, but could one of you experiencing this problem also test out the fix in #1280 and let me know if it resolves the issue for you too?

**Comment by Wildcarde** [comment]
Unfortunately I won't be able to test this until the end of the semester, we try not to touch the system at all once classes have started unless we absolutely have to.
